### DISCUSSION TITLE
Lego 42160 Audi RS e-tron custom control with xbox controller

### DISCUSSION THREAD

**Discussion by vanczakp** [pybricks/support#1533] [Show and tell]
Hi everyone, here is my second iteration of the RS e-tron control with xbox controller. Feel free to try it and use it.
Detials:

**- Control**
   **Pedal**:  left and right trigger, Right as forward direction, left is backwards direction
   **Steering**: left analog horizontal.
   **X**: Drive with power control which is realistic and can be configured as you want (default when startup)
   **B**: Drive with direct speed control which means the imput parameter from trigger is the target speed (0-100) without limitation of 
       acceleration. Use it for precise slow manouvering.

**-Drive modes**
   **Power control** - realistic: when you press the gas to go forward, you will accelerate with the acceleration parameter. If release a bit the 
   trigger comes the motor break which uses the acceleration parameter currently.
   When you completely release the trigger then comes the motor break like run out which can be controlled by the nogas parameter
   When the car drives at a certain direction, and you press the opposite direction trigger comes the break. Break force depends on how 
   hard you press the trigger. After the car has stopped it will sart to accelerate to backwards.  Works the same while backwards.

  **Drive with direct speed control** which means the imput parameter from trigger is the target speed (0-100) without limitation of 
  acceleration. Use it for precise slow manouvering. It acts like you have a differential gear lock between the 
  two motors.

@pybricks you can add this as example later, but first I would like to update it to use only different acceleration instead of ramping (which I created originally for powered up app) and will add a third mode with realistic speed control.

Create a new file with coding layout and copy the following text and paste into it.

```python
#created by Peter Vanczak
#Version 02
from pybricks.hubs import TechnicHub
from pybricks.iodevices import XboxController
from pybricks.parameters import Button, Color, Direction, Port
from pybricks.pupdevices import Motor
from pybricks.robotics import Car
from pybricks.tools import wait

# Set up all devices.
Audi_RS_e_tron = TechnicHub()
front = Motor(Port.A, Direction.CLOCKWISE)
rear = Motor(Port.B, Direction.CLOCKWISE)
steer = Motor(Port.D, Direction.CLOCKWISE)
car = Car(steer, [front, rear])
controller = XboxController()

# Initialize variables.
control_profile_actual = 0
Control_porfile_choosen = 0
max_speed = 1500
pedal = 0
power = 0
acceleration = 2000 / 1000
nogas = 200 / 1000
brake = 4000 / 1000
motorbrake = 1000 / 1000

def direct_speed_control():
    car.drive_speed((max_speed / 100) * pedal)

def power_mode_realistic():
    global power
    if pedal == 0:
        if abs(power) <= 2:
            power = 0
        else:
            if power > 0:
                power = power - nogas
            else:
                power = power + nogas
    else:
        if power == 0:
            power = (pedal / abs(pedal)) * acceleration
        else:
            if power > 0:
                if pedal > 0:
                    if abs(power - pedal) <= acceleration * 0.66:
                        power = power
                    else:
                        if pedal > power:
                            power = power + acceleration
                        else:
                            power = power - motorbrake
                else:
                    power = power - brake * abs(pedal / 100)
                    if power <= 2:
                        power = 0
            else:
                if pedal < 0:
                    if abs(power - pedal) <= acceleration * 0.66:
                        power = power
                    else:
                        if pedal > power:
                            power = power + motorbrake
                        else:
                            power = power - acceleration
                else:
                    power = power + brake * abs(pedal / 100)
                    if power >= -2:
                        power = 0
    car.drive_power(power)


# The main program starts here.
front.control.limits(speed=max_speed)
rear.control.limits(speed=max_speed)
Audi_RS_e_tron.light.on(Color.BLUE)
while True:
    if Button.X in controller.buttons.pressed():
        Control_porfile_choosen = 0
        Audi_RS_e_tron.light.on(Color.BLUE)
    else:
        pass
    if Button.B in controller.buttons.pressed():
        Control_porfile_choosen = 1
        Audi_RS_e_tron.light.on(Color.MAGENTA)
    else:
        pass
    if Control_porfile_choosen != control_profile_actual:
        if Control_porfile_choosen == 0:
            front.control.pid(42484, 21242, 5310, 8, 15)
            rear.control.pid(42484, 21242, 5310, 8, 15)
        else:
            front.control.pid(10000, 21242, 5310, 8, 15)
            rear.control.pid(10000, 21242, 5310, 8, 15)
        control_profile_actual = Control_porfile_choosen
    else:
        pass
    pedal = controller.triggers()[1] - controller.triggers()[0]
    if control_profile_actual == 0:
        power_mode_realistic()
    else:
        direct_speed_control()
    car.steer(controller.joystick_left()[0])
    wait(10)
```

**Comment by johnmrichey** [comment]
I am trying to get my xbox controller to connect to my Hub running the latest beta firmware of pybricks and I have checked that my xbox controller is up to date using the Xbox Accessories app.  But for whatever reason the Hub never connects to the Xbox Controller.

Any help would be appreciated.

It's my understanding there is a FW 5.17 version available but when I check the FW over USB it shows as 5.14.  I have double checked in the Xbox Accessories app and it says the controller is up to date.

**Reply by vanczakp** [reply]
To be honest it is not that trivial how it connects. When first connect (or when it was connected to other device) put it pairing mode right before you start the program. I observed after two seconds it wont connect. When hub was the last connected device, just turn on the controller right before you start the program. It does not matter if you start the program on the hub or via pybricks.

**Comment by vasilegh** [comment]
Hello,
Is there a way to configure PORT C to run the lights? I tried coding, but after I run the program the controller just disconnects.

**Reply by vanczakp** [reply]
Hi! Ask the developers. Open a new topic here somewhere.
I never tried using lights but what you described is your program stops when you try to turn on the light…
